                                        ## lectura de JSON y cracion de DataFrames ##

Importamos librerias para trabajar

In [158]:
import json
import pandas as pd
import numpy as np
import os
import ast

Ruta relativa

In [159]:
# creamos las variables para las Rutas relativas a los archivos JSON

ruta_AUS_User_Reviews = os.path.join('JSON', 'australian_user_reviews.json')

ruta_AUS_User_Items = os.path.join('JSON', 'australian_users_items.json')

ruta_Output_Steam_Games = os.path.join('JSON', 'output_steam_games.json')

Leemos los archivos .JSON y los cargamos a un Dataframe

In [160]:
# vamos a leer el contenido del archivo JSON (australian_user_reviews.json)
with open(ruta_AUS_User_Reviews, 'r', encoding='utf-8') as f:
    data = f.readlines()

# vamos a convertir las lineas en registros
records = [eval(line.strip()) for line in data]

# Creamos el DataFrame a partir de los records o registros
df_UserReviews= pd.DataFrame(records)

In [161]:
# vamos a leer el contenido del archivo JSON (australian_users_items.json)
rows = []
with open (ruta_AUS_User_Items, encoding='MacRoman') as f:
    for line in f.readlines():
        rows.append(ast.literal_eval(line))

# Creamos el DataFrame a partir de los records o registros
df_UserItems = pd.DataFrame(rows)

In [162]:
# vamos a leer el contenido del archivo JSON (output_steam_games.json)

# Creamos el DataFrame a partir leer el .json
df_OutputSteamGames=pd.read_json(ruta_Output_Steam_Games,lines=True, encoding='MacRoman')

Chequeamos los dataframes creados anteriormente y procesedemos a desanidar las columnas anidadas

                                                df_UserReviews

In [163]:
# Podemos verificar la estructura y tipo de dato del dataframe
df_UserReviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25799 entries, 0 to 25798
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   user_id   25799 non-null  object
 1   user_url  25799 non-null  object
 2   reviews   25799 non-null  object
dtypes: object(3)
memory usage: 604.8+ KB


In [164]:
# Chequeamos como quedo el Dataframe df_UserReviews
df_UserReviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                                             reviews  
0  [{'funny': '', 'posted': 'Posted November 5, 2...  
1  [{'funny': '', 'posted': 'Posted June 24, 2014...  
2  [{'funny': '', 'posted': 'Posted February 3.',...  
3  [{'funny': '', 'posted': 'Posted October 14, 2...  
4  [{'funny': '3 people found this review funny',...

Como podemos ver en el dataframe, en la columna reviews, los datos se encuentran anidados, por ende procederemos a desanidar los datos y cargarlos en el dataframe

In [165]:
# Desanidamos la columna review
desanidado = []
for index, row in df_UserReviews.iterrows():
    user_id = row['user_id']
    user_url = row['user_url']
    for review_dict in row['reviews']:
        review_dict['user_id'] = user_id
        review_dict['user_url'] = user_url
        desanidado.append(review_dict)

df_UserReviews = pd.DataFrame(desanidado)

In [166]:
# Volvemos a chequear el dataframe df_UserReviews
df_UserReviews.head()

funny                     posted last_edited item_id  \
0         Posted November 5, 2011.                1250   
1            Posted July 15, 2011.               22200   
2           Posted April 21, 2011.               43110   
3            Posted June 24, 2014.              251610   
4        Posted September 8, 2013.              227300   

                                           helpful  recommend  \
0                                   No ratings yet       True   
1                                   No ratings yet       True   
2                                   No ratings yet       True   
3  15 of 20 people (75%) found this review helpful       True   
4     0 of 1 people (0%) found this review helpful       True   

                                              review            user_id  \
0  Simple yet with great replayability. In my opi...  76561197970982479   
1               It's unique and worth a playthrough.  76561197970982479   
2  Great atmosphere. The gunplay can be a bit chu...  76561197970982479   
3  I know what you think when you see this title ...            js41637   
4  For a simple (it's actually not all that simpl...            js41637   

                                            user_url  
0  http://steamcommunity.com/profiles/76561197970...  
1  http://steamcommunity.com/profiles/76561197970...  
2  http://steamcommunity.com/profiles/76561197970...  
3               http://steamcommunity.com/id/js41637  
4               http://steamcommunity.com/id/js41637

                                            df_UserItems

In [167]:
df_UserItems.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88310 entries, 0 to 88309
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      88310 non-null  object
 1   items_count  88310 non-null  int64 
 2   steam_id     88310 non-null  object
 3   user_url     88310 non-null  object
 4   items        88310 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.4+ MB


In [168]:
# Chequeamos como quedo el Dataframe df_UserItems
df_UserItems.head()

user_id  items_count           steam_id  \
0  76561197970982479          277  76561197970982479   
1            js41637          888  76561198035864385   
2          evcentric          137  76561198007712555   
3         Riot-Punch          328  76561197963445855   
4              doctr          541  76561198002099482   

                                            user_url  \
0  http://steamcommunity.com/profiles/76561197970...   
1               http://steamcommunity.com/id/js41637   
2             http://steamcommunity.com/id/evcentric   
3            http://steamcommunity.com/id/Riot-Punch   
4                 http://steamcommunity.com/id/doctr   

                                               items  
0  [{'item_id': '10', 'item_name': 'Counter-Strik...  
1  [{'item_id': '10', 'item_name': 'Counter-Strik...  
2  [{'item_id': '1200', 'item_name': 'Red Orchest...  
3  [{'item_id': '10', 'item_name': 'Counter-Strik...  
4  [{'item_id': '300', 'item_name': 'Day of Defea...

Como se ve en este Dataframe tambien estan anidados datos en la columna items. procedemos a desanidar la misma

In [169]:
# Desanidamos la columna items

desanidado = []
for index, row in df_UserItems.iterrows():
    user_id = row['user_id']
    user_url = row['user_url']
    user_items_count = row['items_count']
    for df_items_dict in row['items']:
        df_items_dict['user_id'] = user_id
        df_items_dict['user_url'] = user_url
        df_items_dict['items_count'] = user_items_count
        desanidado.append(df_items_dict)

df_UserItems = pd.DataFrame(desanidado)

In [170]:
# Volvemos a chequear el dataframe df_UserItems
df_UserItems.head()

item_id                  item_name  playtime_forever  playtime_2weeks  \
0      10             Counter-Strike                 6                0   
1      20      Team Fortress Classic                 0                0   
2      30              Day of Defeat                 7                0   
3      40         Deathmatch Classic                 0                0   
4      50  Half-Life: Opposing Force                 0                0   

             user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
4  76561197970982479  http://steamcommunity.com/profiles/76561197970...   

   items_count  
0          277  
1          277  
2          277  
3          277  
4          277

                                            df_OutputSteamGames

In [171]:
df_OutputSteamGames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   publisher     24083 non-null  object 
 1   genres        28852 non-null  object 
 2   app_name      32133 non-null  object 
 3   title         30085 non-null  object 
 4   url           32135 non-null  object 
 5   release_date  30068 non-null  object 
 6   tags          31972 non-null  object 
 7   reviews_url   32133 non-null  object 
 8   specs         31465 non-null  object 
 9   price         30758 non-null  object 
 10  early_access  32135 non-null  float64
 11  id            32133 non-null  float64
 12  developer     28836 non-null  object 
dtypes: float64(2), object(11)
memory usage: 11.9+ MB


In [172]:
# Chequeamos como quedo el Dataframe df_OutputSteamGames
df_OutputSteamGames.head()

publisher genres app_name title   url release_date  tags reviews_url specs  \
0      None   None     None  None  None         None  None        None  None   
1      None   None     None  None  None         None  None        None  None   
2      None   None     None  None  None         None  None        None  None   
3      None   None     None  None  None         None  None        None  None   
4      None   None     None  None  None         None  None        None  None   

  price  early_access  id developer  
0  None           NaN NaN      None  
1  None           NaN NaN      None  
2  None           NaN NaN      None  
3  None           NaN NaN      None  
4  None           NaN NaN      None

In [173]:
# Chequeamos como quedo el Dataframe df_OutputSteamGames los registros del final del df
df_OutputSteamGames.tail()

publisher                                 genres  \
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444             None                                   None   

                        app_name                     title  \
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                      None   

                                                      url release_date  \
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...         None   

                                                     tags  \
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price  early_access  \
120440                [Single-player, Steam Achievements]  1.99           0.0   
120441  [Single-player, Steam Achievements, Steam Clou...  4.99           0.0   
120442  [Single-player, Steam Achievements, Steam Trad...  1.99           0.0   
120443   [Single-player, Steam Achievements, Steam Cloud]  4.99           0.0   
120444  [Single-player, Stats, Steam Leaderboards, HTC...  4.99           1.0   

              id                 developer  
120440  773640.0        Nikita "Ghost_RUS"  
120441  733530.0                    Sacada  
120442  610660.0  Laush Dmitriy Sergeevich  
120443  658870.0             xropi,stev3ns  
120444  681550.0                      None

                                    Proceso de Transformacion de datos

En este paso empezaremos la transformación de datos.
En esta etapa vamos a evaluar el Dataframe y analizar que columnas son relevantes, tambien vamos a chequear datos nulos y rellenar filas con nulos para los campos que lo precisan.

Empezamos a analizar el Dataframe df_UserItems

In [174]:
# Vizualizamos el dataframe para evaluar que columnas son utiles
df_UserItems.sort_values(by='playtime_forever', ascending=False).head()

item_id                              item_name  playtime_forever  \
587715     4000                            Garry's Mod            642773   
2499068  212200                               Mabinogi            635295   
4075729   42710  Call of Duty: Black Ops - Multiplayer            632452   
1495340    4000                            Garry's Mod            613411   
1836985   72200                       Universe Sandbox            600068   

         playtime_2weeks            user_id  \
587715              7299              wolop   
2499068                0        Evilutional   
4075729             8126  76561198019826668   
1495340              588  76561198039832932   
1836985                0         tsunamitad   

                                                  user_url  items_count  
587715                  http://steamcommunity.com/id/wolop          107  
2499068           http://steamcommunity.com/id/Evilutional           78  
4075729  http://steamcommunity.com/profiles/76561198019...           12  
1495340  http://steamcommunity.com/profiles/76561198039...           90  
1836985            http://steamcommunity.com/id/tsunamitad          793

Eliminar Columnas No Necesarias

In [175]:
# Eliminamos las columnas que no son utiles para el desarrollo del proyecto
df_UserItems = df_UserItems.drop(['user_url', 'playtime_2weeks'], axis=1)

In [176]:
# Chequeamos como quedo df_UserItems
df_UserItems.head()

item_id                  item_name  playtime_forever            user_id  \
0      10             Counter-Strike                 6  76561197970982479   
1      20      Team Fortress Classic                 0  76561197970982479   
2      30              Day of Defeat                 7  76561197970982479   
3      40         Deathmatch Classic                 0  76561197970982479   
4      50  Half-Life: Opposing Force                 0  76561197970982479   

   items_count  
0          277  
1          277  
2          277  
3          277  
4          277

Verificamos los valores nulos del Dataframe

In [177]:
# Verificamos valores nulos en cada columna
valores_nulos_por_columna = df_UserItems.isnull().sum()

# imprimimos los resultados
print(valores_nulos_por_columna)

item_id             0
item_name           0
playtime_forever    0
user_id             0
items_count         0
dtype: int64


Verificamos los valores duplicados

In [178]:
# Verificamos filas duplicadas en el dataframe
duplicados = df_UserItems.duplicated()

# Mostramos los duplicados
print(df_UserItems[duplicados])

        item_id                             item_name  playtime_forever  \
164294       20                 Team Fortress Classic                 5   
164295       50             Half-Life: Opposing Force                 0   
164296       70                             Half-Life                 0   
164297      130                 Half-Life: Blue Shift                 0   
164298      220                           Half-Life 2               198   
...         ...                                   ...               ...   
4898223  213670  South Park‚Ñ¢: The Stick of Truth‚Ñ¢               725   
4898224  221910                   The Stanley Parable                53   
4898225  261030          The Walking Dead: Season Two               253   
4898226  273110         Counter-Strike Nexon: Zombies                 0   
4898227     730      Counter-Strike: Global Offensive                 0   

                   user_id  items_count  
164294              Nikiad          109  
164295         

In [179]:
# Con este codigo obtenemos el número de filas y columnas usando shape
num_filas, num_columnas = df_UserItems.shape
print(f'Número de filas: {num_filas}, Número de columnas: {num_columnas}')

Número de filas: 5153209, Número de columnas: 5


In [180]:
# Procedemos a borrar los duplicados
df_UserItems.drop_duplicates(inplace=True)

In [181]:
# Obtenemos el número de filas y columnas usando shape que nos quedo despues de dropear duplicados
num_filas, num_columnas = df_UserItems.shape
print(f'Número de filas: {num_filas}, Número de columnas: {num_columnas}')

Número de filas: 5094092, Número de columnas: 5


Reorganizar y renombrar columnas

In [182]:
# vamos a renombrar las columnas en una manera standard y siguiendo buenas practicas 
df_UserItems = df_UserItems.rename(columns={'item_id': 'ItemId'})
df_UserItems = df_UserItems.rename(columns={'item_name': 'ItemName'})
df_UserItems = df_UserItems.rename(columns={'playtime_forever': 'PlaytimeForever'})
df_UserItems = df_UserItems.rename(columns={'user_id': 'UserId'})
df_UserItems = df_UserItems.rename(columns={'items_count': 'ItemsCount'})

In [183]:
# chequeamos los cambios en el nombre de la columna y chequeamos los tipos de datos de las columnas
df_UserItems.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5094092 entries, 0 to 5153208
Data columns (total 5 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   ItemId           object
 1   ItemName         object
 2   PlaytimeForever  int64 
 3   UserId           object
 4   ItemsCount       int64 
dtypes: int64(2), object(3)
memory usage: 233.2+ MB


Convertir tipos de datos

In [184]:
# chequeamos los tipos de datos del dataframe
df_UserItems.dtypes

ItemId             object
ItemName           object
PlaytimeForever     int64
UserId             object
ItemsCount          int64
dtype: object

Corremos el siguiente codigo para chequear si los valores de la columna ItemId son enteros ya que son los ids de los titulos y considero importante chequear

In [185]:
# 'ItemId' es la columna que deseas verificar
df_UserItems['ItemId'] = pd.to_numeric(df_UserItems['ItemId'], errors='coerce')

# Filtramos las filas que no son enteros
es_entero = df_UserItems['ItemId'].notnull() & (df_UserItems['ItemId'] % 1 == 0)
resultados = df_UserItems[~es_entero]

# Mostrar los resultados
print(resultados)

Empty DataFrame
Columns: [ItemId, ItemName, PlaytimeForever, UserId, ItemsCount]
Index: []


un DataFrame vacío, indica que no se encontraron filas que cumplieran con la condición especificada en tu filtro. En otras palabras, todos los valores en la columna 'ItemId' parecen ser numéricos (o NaN después de la conversión) y cumplen con la condición de ser enteros.

In [186]:
# cambiamos el tipo de dato de la columna ItemId a int
df_UserItems['ItemId'] = df_UserItems['ItemId'].astype(int)

In [187]:
# cambiamos el tipo de dato de la columna ItemsCount a int
df_UserItems['ItemsCount'] = df_UserItems['ItemsCount'].astype(int)

In [188]:
# cambiamos el tipo de dato de la columna PlaytimeForever a int
df_UserItems['PlaytimeForever'] = df_UserItems['PlaytimeForever'].astype(int)

In [189]:
# cambiamos el tipo de dato de la columna UserId a str
df_UserItems['UserId'] = df_UserItems['UserId'].astype(str)

In [190]:
# cambiamos el tipo de dato de la columna ItemName a str
df_UserItems['ItemName'] = df_UserItems['ItemName'].astype(str)

**Observacion:**

En Python, el tipo de datos de texto se conoce como secuencia de caracteres (string). En Pandas se los conoce como objetos (object). Las secuencias de caracteres pueden contener números y / o caracteres.

chequeamos como nos quedaron los cambios de tipo de datos

In [191]:
# chequeamos los tipos de datos del dataframe
tipos_de_datos = df_UserItems.dtypes
print(tipos_de_datos)

ItemId              int32
ItemName           object
PlaytimeForever     int32
UserId             object
ItemsCount          int32
dtype: object


In [192]:
# chequeamos como nos quedo el dataframe df_UserItems con los procesos de transformacion realizados
df_UserItems.head()

ItemId                   ItemName  PlaytimeForever             UserId  \
0      10             Counter-Strike                6  76561197970982479   
1      20      Team Fortress Classic                0  76561197970982479   
2      30              Day of Defeat                7  76561197970982479   
3      40         Deathmatch Classic                0  76561197970982479   
4      50  Half-Life: Opposing Force                0  76561197970982479   

   ItemsCount  
0         277  
1         277  
2         277  
3         277  
4         277

**Aclaracion:**

la columna PlaytimeForever representa la cantidad de MINUTOS que el usuario juego ese juego desde que lo compro y descargo.
esta representado en minutos, pero divinedo la columna sobre 60 , obtenemos la cantidad de horas jugadas totales de ese usuario en ese juego

Ahora guardamos el Dataframe en un .csv ya con los procesos de transformacion realizados

In [193]:
# Definimos la ruta relativa
ruta_relativa = os.path.join('Datasets-Limpios', 'user_items.parquet')

# Guardamos el DataFrame en un archivo CSV con la ruta relativa
df_UserItems.to_parquet(ruta_relativa, index=False)

-----------------------------------------------------------------------------------------------------------------------

Analizamos el Dataframe df_OutputSteamGames

In [194]:
# chequeamos el dataset df_outputSteamgames
df_OutputSteamGames.head()

publisher genres app_name title   url release_date  tags reviews_url specs  \
0      None   None     None  None  None         None  None        None  None   
1      None   None     None  None  None         None  None        None  None   
2      None   None     None  None  None         None  None        None  None   
3      None   None     None  None  None         None  None        None  None   
4      None   None     None  None  None         None  None        None  None   

  price  early_access  id developer  
0  None           NaN NaN      None  
1  None           NaN NaN      None  
2  None           NaN NaN      None  
3  None           NaN NaN      None  
4  None           NaN NaN      None

Verificamos que varias filas no contienen datos y hay que eliminarlas

In [195]:
# Borramos las filas donde todos los valores son None o nan
df_OutputSteamGames.dropna(how='all', inplace=True)

In [196]:
# Chequeamos el dataframe para ver como quedo
df_OutputSteamGames.head()

publisher                                             genres  \
88310         Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
88311  Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
88312      Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
88313              彼岸领域                        [Action, Adventure, Casual]   
88314              None                                               None   

                      app_name                    title  \
88310      Lost Summoner Kitty      Lost Summoner Kitty   
88311                Ironbound                Ironbound   
88312  Real Pool 3D - Poolians  Real Pool 3D - Poolians   
88313                  弹炸人2222                  弹炸人2222   
88314            Log Challenge                     None   

                                                     url release_date  \
88310  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
88311  http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
88312  http://store.steampowered.com/app/670290/Real_...   2017-07-24   
88313     http://store.steampowered.com/app/767400/2222/   2017-12-07   
88314  http://store.steampowered.com/app/773570/Log_C...         None   

                                                    tags  \
88310      [Strategy, Action, Indie, Casual, Simulation]   
88311  [Free to Play, Strategy, Indie, RPG, Card Game...   
88312  [Free to Play, Simulation, Sports, Casual, Ind...   
88313                        [Action, Adventure, Casual]   
88314                    [Action, Indie, Casual, Sports]   

                                             reviews_url  \
88310  http://steamcommunity.com/app/761140/reviews/?...   
88311  http://steamcommunity.com/app/643980/reviews/?...   
88312  http://steamcommunity.com/app/670290/reviews/?...   
88313  http://steamcommunity.com/app/767400/reviews/?...   
88314  http://steamcommunity.com/app/773570/reviews/?...   

                                                   specs         price  \
88310                                    [Single-player]          4.99   
88311  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
88312  [Single-player, Multi-player, Online Multi-Pla...  Free to Play   
88313                                    [Single-player]          0.99   
88314  [Single-player, Full controller support, HTC V...          2.99   

       early_access        id         developer  
88310           0.0  761140.0         Kotoshiro  
88311           0.0  643980.0  Secret Level SRL  
88312           0.0  670290.0      Poolians.com  
88313           0.0  767400.0              彼岸领域  
88314           0.0  773570.0              None

Verificamos los valores duplicados dentro de df_OutptSteamGames

In [197]:
# Obtenemos el número de filas y columnas usando shape
num_filas, num_columnas = df_OutputSteamGames.shape
print(f'Número de filas: {num_filas}, Número de columnas: {num_columnas}')

Número de filas: 32135, Número de columnas: 13


In [198]:
df_OutputSteamGames.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'early_access', 'id',
       'developer'],
      dtype='object')

In [199]:
# Verificamos filas duplicadas en el dataframe
columnas_a_verificar = ['publisher', 'app_name', 'title', 'url', 'release_date',
                        'reviews_url', 'price', 'early_access', 'id',
                        'developer']
duplicados_en_columnas = df_OutputSteamGames.duplicated(subset=columnas_a_verificar)

# Mostramos los duplicados
print(df_OutputSteamGames[duplicados_en_columnas])

Empty DataFrame
Columns: [publisher, genres, app_name, title, url, release_date, tags, reviews_url, specs, price, early_access, id, developer]
Index: []


Verificamos sacando las listas que no hay valores duplicados

Vemos los valores nulos del Dataframe

In [200]:
# Chequemos el dataframe para ver la cantidad de valores nulos
df_OutputSteamGames.isnull().sum()

publisher       8052
genres          3283
app_name           2
title           2050
url                0
release_date    2067
tags             163
reviews_url        2
specs            670
price           1377
early_access       0
id                 2
developer       3299
dtype: int64

In [201]:
len(df_OutputSteamGames)

32135

**Nota:** 

Podemos ver en el campo 'genres' los valores faltantes son 3283, pero la informacion que falta podriamos sacarla del campo 'tags', solo que habria que obtener de ese campo los datos del genero y no el resto. Esa podria ser una buena solucion al problema.

Vemos que en el campo app_name y title contienen una informacion redundante, ya que los dos campos describen el titulo de la aplicacion solo que en dos formatos diferentes.
En este caso podemos utilizar el campo app_name para identificar el titulo del producto porque como vemos en los datos nulos o faltantes en ese campo solo tiene 2 faltantes.

Hay que chequear el formato de la fecha de release_date

el Campo 'developer' es importante conteniendo solo 3299 faltantes de 32135 totales.

Eliminar Columnas No Necesarias

In [202]:
df_OutputSteamGames.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'early_access', 'id',
       'developer'],
      dtype='object')

In [203]:
# Eliminamos las columnas que no son utiles para el desarrollo del proyecto
df_OutputSteamGames.drop(columns=["publisher","title","url","early_access","reviews_url","specs"], inplace=True)

In [204]:
# Chequeamos como quedo el Dataframe
df_OutputSteamGames.head()

genres  \
88310      [Action, Casual, Indie, Simulation, Strategy]   
88311               [Free to Play, Indie, RPG, Strategy]   
88312  [Casual, Free to Play, Indie, Simulation, Sports]   
88313                        [Action, Adventure, Casual]   
88314                                               None   

                      app_name release_date  \
88310      Lost Summoner Kitty   2018-01-04   
88311                Ironbound   2018-01-04   
88312  Real Pool 3D - Poolians   2017-07-24   
88313                  弹炸人2222   2017-12-07   
88314            Log Challenge         None   

                                                    tags         price  \
88310      [Strategy, Action, Indie, Casual, Simulation]          4.99   
88311  [Free to Play, Strategy, Indie, RPG, Card Game...  Free To Play   
88312  [Free to Play, Simulation, Sports, Casual, Ind...  Free to Play   
88313                        [Action, Adventure, Casual]          0.99   
88314                    [Action, Indie, Casual, Sports]          2.99   

             id         developer  
88310  761140.0         Kotoshiro  
88311  643980.0  Secret Level SRL  
88312  670290.0      Poolians.com  
88313  767400.0              彼岸领域  
88314  773570.0              None

Reorganizamos y renombramos las columnas

In [205]:
# chequeamos la columnas nombradas anteriormente para df_UserItems
df_UserItems.columns

Index(['ItemId', 'ItemName', 'PlaytimeForever', 'UserId', 'ItemsCount'], dtype='object')

In [206]:
#chequeamos los nombres de las columnas
df_OutputSteamGames.columns

Index(['genres', 'app_name', 'release_date', 'tags', 'price', 'id',
       'developer'],
      dtype='object')

In [207]:
# vamos a renombrar las columnas en una manera standard y siguiendo buenas practicas 
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'genres': 'Genres'}) 
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'app_name': 'Title'})
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'release_date': 'ReleaseDate'})
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'price': 'Price'})
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'id': 'ItemId'})
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'developer': 'Developer'})
df_OutputSteamGames = df_OutputSteamGames.rename(columns={'tags': 'Tags'})


In [208]:
# chequeamos los cambios de nombres de columnas
df_OutputSteamGames.head(1)

Genres                Title  \
88310  [Action, Casual, Indie, Simulation, Strategy]  Lost Summoner Kitty   

      ReleaseDate                                           Tags Price  \
88310  2018-01-04  [Strategy, Action, Indie, Casual, Simulation]  4.99   

         ItemId  Developer  
88310  761140.0  Kotoshiro

verificamos el tipo de dato de las columnas

In [209]:
# chequeamos los tipos de datos del dataframe
df_OutputSteamGames.dtypes

Genres          object
Title           object
ReleaseDate     object
Tags            object
Price           object
ItemId         float64
Developer       object
dtype: object

Como vemos para los campos Genres, Title, Tags y Developer los tipos de datos quedaran en tipo object

In [210]:
df_OutputSteamGames.head(2)

Genres                Title  \
88310  [Action, Casual, Indie, Simulation, Strategy]  Lost Summoner Kitty   
88311           [Free to Play, Indie, RPG, Strategy]            Ironbound   

      ReleaseDate                                               Tags  \
88310  2018-01-04      [Strategy, Action, Indie, Casual, Simulation]   
88311  2018-01-04  [Free to Play, Strategy, Indie, RPG, Card Game...   

              Price    ItemId         Developer  
88310          4.99  761140.0         Kotoshiro  
88311  Free To Play  643980.0  Secret Level SRL

Cambiamos los tipos de datos de las columnas

Para cambiar los siguientes tipos de datos de las columnas, tenemos que revisar los datos que contienen

In [211]:
# chequeamos que datos contiene la columna ReleaseDate
print(df_OutputSteamGames['ReleaseDate'].unique())

['2018-01-04' '2017-07-24' '2017-12-07' ... '2016-11-19' 'January 2018'
 '2018-10-01']


In [212]:
# cambiamos el tipo de dato de la columna ReleaseDate a datetime format
df_OutputSteamGames['ReleaseDate'] = pd.to_datetime(df_OutputSteamGames['ReleaseDate'], errors='coerce', format='%Y-%m-%d', exact=False)

In [213]:
# Confirmamos el cambio de formato
print(df_OutputSteamGames['ReleaseDate'].unique())

['2018-01-04T00:00:00.000000000' '2017-07-24T00:00:00.000000000'
 '2017-12-07T00:00:00.000000000' ... '2004-03-01T00:00:00.000000000'
 '2016-11-19T00:00:00.000000000' '2018-10-01T00:00:00.000000000']


In [214]:
df_OutputSteamGames['ReleaseDate']

88310    2018-01-04
88311    2018-01-04
88312    2017-07-24
88313    2017-12-07
88314           NaT
            ...    
120440   2018-01-04
120441   2018-01-04
120442   2018-01-04
120443   2017-09-02
120444          NaT
Name: ReleaseDate, Length: 32135, dtype: datetime64[ns]

In [215]:
# chequeamos que datos contiene la columna Price
print(df_OutputSteamGames['Price'].unique())

[4.99 'Free To Play' 'Free to Play' 0.99 2.99 3.99 9.99 18.99 29.99 None
 'Free' 10.99 1.5899999999999999 14.99 1.99 59.99 8.99 6.99 7.99 39.99
 19.99 7.49 12.99 5.99 2.49 15.99 1.25 24.99 17.99 61.99 3.49 11.99 13.99
 'Free Demo' 'Play for Free!' 34.99 74.76 1.49 32.99 99.99 14.95 69.99
 16.99 79.99 49.99 5.0 44.99 13.98 29.96 119.99 109.99 149.99 771.71
 'Install Now' 21.99 89.99 'Play WARMACHINE: Tactics Demo' 0.98 139.92
 4.29 64.99 'Free Mod' 54.99 74.99 'Install Theme' 0.89 'Third-party' 0.5
 'Play Now' 299.99 1.29 3.0 15.0 5.49 23.99 49.0 20.99 10.93
 1.3900000000000001 'Free HITMAN™ Holiday Pack' 36.99 4.49 2.0 4.0 9.0
 234.99 1.9500000000000002 1.5 199.0 189.0 6.66 27.99 10.49 129.99 179.0
 26.99 399.99 31.99 399.0 20.0 40.0 3.33 199.99 22.99 320.0 38.85 71.7
 59.95 995.0 27.49 3.39 6.0 19.95 499.99 16.06 4.68 131.4 44.98 202.76 1.0
 2.3 0.9500000000000001 172.24 249.99 2.9699999999999998 10.96 10.0 30.0
 2.66 6.48 19.29 11.15 18.9 2.89 'Play the Demo' 99.0 87.94 599.0 8.98
 9

Vemos que en la columna de Price aparecen valores string(como free to play o starting at $499), lo quye podemos hacer es cambiarlos a 0

In [216]:
# cambiamos los valores de string a int y dejamos nan los faltantes
df_OutputSteamGames['Price'] = df_OutputSteamGames['Price'].apply(lambda x: 0 if pd.notna(x) and not isinstance(x, (int, float)) else x).astype(float)

In [217]:
# chequeamos que datos contiene la columna Price
print(df_OutputSteamGames['Price'].unique())

[4.9900e+00 0.0000e+00 9.9000e-01 2.9900e+00 3.9900e+00 9.9900e+00
 1.8990e+01 2.9990e+01        nan 1.0990e+01 1.5900e+00 1.4990e+01
 1.9900e+00 5.9990e+01 8.9900e+00 6.9900e+00 7.9900e+00 3.9990e+01
 1.9990e+01 7.4900e+00 1.2990e+01 5.9900e+00 2.4900e+00 1.5990e+01
 1.2500e+00 2.4990e+01 1.7990e+01 6.1990e+01 3.4900e+00 1.1990e+01
 1.3990e+01 3.4990e+01 7.4760e+01 1.4900e+00 3.2990e+01 9.9990e+01
 1.4950e+01 6.9990e+01 1.6990e+01 7.9990e+01 4.9990e+01 5.0000e+00
 4.4990e+01 1.3980e+01 2.9960e+01 1.1999e+02 1.0999e+02 1.4999e+02
 7.7171e+02 2.1990e+01 8.9990e+01 9.8000e-01 1.3992e+02 4.2900e+00
 6.4990e+01 5.4990e+01 7.4990e+01 8.9000e-01 5.0000e-01 2.9999e+02
 1.2900e+00 3.0000e+00 1.5000e+01 5.4900e+00 2.3990e+01 4.9000e+01
 2.0990e+01 1.0930e+01 1.3900e+00 3.6990e+01 4.4900e+00 2.0000e+00
 4.0000e+00 9.0000e+00 2.3499e+02 1.9500e+00 1.5000e+00 1.9900e+02
 1.8900e+02 6.6600e+00 2.7990e+01 1.0490e+01 1.2999e+02 1.7900e+02
 2.6990e+01 3.9999e+02 3.1990e+01 3.9900e+02 2.0000e+01 4.0000

In [218]:
df_OutputSteamGames['Price']

88310     4.99
88311     0.00
88312     0.00
88313     0.99
88314     2.99
          ... 
120440    1.99
120441    4.99
120442    1.99
120443    4.99
120444    4.99
Name: Price, Length: 32135, dtype: float64

In [219]:
# Ahora vamos a cambiar la columna ItemId, para eso vamos a ver que datos contiene

# chequeamos que datos contiene la columna 
print(df_OutputSteamGames['ItemId'].unique())

[761140. 643980. 670290. ... 610660. 658870. 681550.]


In [220]:
# vemos que valores tenemos
df_OutputSteamGames['ItemId'].sort_values()

120416         10.0
120413         20.0
120424         30.0
120418         40.0
120414         50.0
            ...    
118632    2028062.0
118834    2028103.0
89735     2028850.0
88384           NaN
119271          NaN
Name: ItemId, Length: 32135, dtype: float64

In [221]:
# Vemos valores NaN faltantes para el ItemId, los vamos a eliminar son solo dos registros faltantes
df_OutputSteamGames = df_OutputSteamGames.dropna(subset=['ItemId'])

In [222]:
# vemos como quedo con el proceso de eliminacion de esos dos registros 
df_OutputSteamGames['ItemId'].sort_values()

120416         10.0
120413         20.0
120424         30.0
120418         40.0
120414         50.0
            ...    
118903    2028055.0
118999    2028056.0
118632    2028062.0
118834    2028103.0
89735     2028850.0
Name: ItemId, Length: 32133, dtype: float64

In [223]:
# cambiamos el tipo de dato de la columna ItemId a int
df_OutputSteamGames.loc[:, 'ItemId'] = df_OutputSteamGames['ItemId'].astype(int)

C:\Users\frank\AppData\Local\Temp\ipykernel_21320\1349772635.py:2: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  df_OutputSteamGames.loc[:, 'ItemId'] = df_OutputSteamGames['ItemId'].astype(int)


Chequeamos como quedaron los cambios de tipos de datos

In [224]:
# Vemos como nos quedo los cambios de tipos de datos
df_OutputSteamGames.dtypes

Genres                 object
Title                  object
ReleaseDate    datetime64[ns]
Tags                   object
Price                 float64
ItemId                  int32
Developer              object
dtype: object

Tratamos los valores nulos

In [225]:
# Chequeamos las columnas que nos quedaron en el dataframe df_OutputGames
df_OutputSteamGames.columns

Index(['Genres', 'Title', 'ReleaseDate', 'Tags', 'Price', 'ItemId',
       'Developer'],
      dtype='object')

In [226]:
# Con esta funcion podemos ver un informe detallado de como quedaron los tipos de datos de los campos y podemos ver
# la cantidad de nulos que tenemos por cada columna

def verificar_tipo_datos(dfoutput):
    mi_dict = {"nombre_campo": [], "tipo_datos": [], "no_nulos_%": [], "nulos_%": []}
    for columna in dfoutput.columns:
        porcentaje_no_nulos = (dfoutput[columna].count() / len(dfoutput)) * 100
        mi_dict["nombre_campo"].append(columna)
        mi_dict["tipo_datos"].append(dfoutput[columna].apply(type).unique())
        mi_dict["no_nulos_%"].append(round(porcentaje_no_nulos, 2))
        mi_dict["nulos_%"].append(round(100-porcentaje_no_nulos, 2))
    df_info = pd.DataFrame(mi_dict)
    for columna in dfoutput.columns:
        print(columna, " (nulos) = ", dfoutput[columna].isnull().sum())
    print("\nfilas completamente nulas: ", dfoutput.isna().all(axis=1).sum())
    return df_info 
df_info = verificar_tipo_datos(df_OutputSteamGames)
print(df_info)

Genres  (nulos) =  3282
Title  (nulos) =  1
ReleaseDate  (nulos) =  2309
Tags  (nulos) =  162
Price  (nulos) =  1377
ItemId  (nulos) =  0
Developer  (nulos) =  3298

filas completamente nulas:  0
  nombre_campo                                         tipo_datos  no_nulos_%  \
0       Genres               [<class 'list'>, <class 'NoneType'>]       89.79   
1        Title                [<class 'str'>, <class 'NoneType'>]      100.00   
2  ReleaseDate  [<class 'pandas._libs.tslibs.timestamps.Timest...       92.81   
3         Tags               [<class 'list'>, <class 'NoneType'>]       99.50   
4        Price                                  [<class 'float'>]       95.71   
5       ItemId                                    [<class 'int'>]      100.00   
6    Developer                [<class 'str'>, <class 'NoneType'>]       89.74   

   nulos_%  
0    10.21  
1     0.00  
2     7.19  
3     0.50  
4     4.29  
5     0.00  
6    10.26  


Genres  (nulos) =  3282

Title  (nulos) =  1

ReleaseDate  (nulos) =  2309

Tags  (nulos) =  162

Price  (nulos) =  1377

ItemId  (nulos) =  0

Developer  (nulos) =  3298


Tendriamos que hacer una ultima transformacion para intentar dejar las columnas lo mejor posible para los proximos pasos

Emepezamos con la columna Title

In [227]:
# Muestra las filas con valores NaN en la columna
missing = df_OutputSteamGames[df_OutputSteamGames['Title'].isna()]

# Muestra las filas con valores NaN en la columna y selecciona solo la columna
missing_values = df_OutputSteamGames.loc[df_OutputSteamGames['Title'].isna(), 'Title']

# Imprime las filas con valores NaN en la columna 
print(missing)

# Imprime los valores específicos NaN en la columna
print(missing_values)

                Genres Title ReleaseDate             Tags  Price  ItemId  \
90890  [Action, Indie]  None  2014-08-26  [Action, Indie]    NaN  317160   

      Developer  
90890      None  
90890    None
Name: Title, dtype: object


Decidi borrar los valores que tiene faltante para la columna Title

In [228]:
# Vemos valor NaN faltante para el campo Title, lo vamos a eliminar
df_OutputSteamGames = df_OutputSteamGames.dropna(subset=['Title'])

In [229]:
# Chequemos ahora como quedo
missing_titles = df_OutputSteamGames[df_OutputSteamGames['Title'].isna()]
print(missing_titles)

Empty DataFrame
Columns: [Genres, Title, ReleaseDate, Tags, Price, ItemId, Developer]
Index: []


Seguimos con la columna Price

podemos ver que hay valores nulos (Price(nulos)=1377)

In [230]:
# Vamos a tratarlos con un 0 en los valores nan
df_OutputSteamGames['Price'].fillna(0, inplace=True)

In [231]:
# chequeamos que hayan sido reemplazados
print(df_OutputSteamGames['Price'].isna().sum())

0


Seguimos con la columna ReleaseDate

In [232]:
# Analizamos como quedaron los valores nan en el campo releasedate, con formato tipo pandas datetime
print(df_OutputSteamGames['ReleaseDate'].unique())

['2018-01-04T00:00:00.000000000' '2017-07-24T00:00:00.000000000'
 '2017-12-07T00:00:00.000000000' ... '2004-03-01T00:00:00.000000000'
 '2016-11-19T00:00:00.000000000' '2018-10-01T00:00:00.000000000']


Ahora vemos la columna Genres

In [233]:
# creamos una lista con los generos existentes
generos=[]
for a in df_OutputSteamGames.Genres.dropna().drop_duplicates():
    for b in a:
        if b not in generos:
            generos.append(b)
generos=sorted(generos)
print(generos)

['Accounting', 'Action', 'Adventure', 'Animation &amp; Modeling', 'Audio Production', 'Casual', 'Design &amp; Illustration', 'Early Access', 'Education', 'Free to Play', 'Indie', 'Massively Multiplayer', 'Photo Editing', 'RPG', 'Racing', 'Simulation', 'Software Training', 'Sports', 'Strategy', 'Utilities', 'Video Production', 'Web Publishing']


In [234]:
# Vemos la cantidad de generos unicos que tenemos
len(generos)

22

In [235]:
#Eliminamos de 'tags' los valores que no son generos
df_OutputSteamGames['Tags']=df_OutputSteamGames['Tags'].dropna().apply(lambda lista: [g for g in lista if g in generos])
#Usamos dropna() para evaluar sobre filas no nulas y apply() para aplicar mi funcion a cada fila
df_OutputSteamGames.loc[df_OutputSteamGames['Genres'].isna() & df_OutputSteamGames['Tags'].notnull(), 'Genres']=df_OutputSteamGames['Tags']

In [236]:
# Mostramos cuantos valores nos quedaron nulos
df_OutputSteamGames['Genres'].isnull().sum()

138

Como vimos con este codigo logramos llevar los faltantes de la columna Genres de 3282 a solo 138

Seguimos con la columna Developer

In [237]:
# Chequeamos nuevamente cuantos faltantes hay en esta columna
df_OutputSteamGames['Developer'].isnull().sum()

3297

In [238]:
# Reemplazarlos los valoresfaltantes con 'Desconocido'
df_OutputSteamGames['Developer'].fillna('Desconocido', inplace=True)

In [239]:
# Chequeamos nuevamente cuantos faltantes hay en esta columna
df_OutputSteamGames['Developer'].isnull().sum()

0

In [240]:
# Volvemos a verificar como quedaron la columnas y los valores faltantes
df_info = verificar_tipo_datos(df_OutputSteamGames)
print(df_info)

Genres  (nulos) =  138
Title  (nulos) =  0
ReleaseDate  (nulos) =  2309
Tags  (nulos) =  162
Price  (nulos) =  0
ItemId  (nulos) =  0
Developer  (nulos) =  0

filas completamente nulas:  0
  nombre_campo                                         tipo_datos  no_nulos_%  \
0       Genres               [<class 'list'>, <class 'NoneType'>]       99.57   
1        Title                                    [<class 'str'>]      100.00   
2  ReleaseDate  [<class 'pandas._libs.tslibs.timestamps.Timest...       92.81   
3         Tags                  [<class 'list'>, <class 'float'>]       99.50   
4        Price                                  [<class 'float'>]      100.00   
5       ItemId                                    [<class 'int'>]      100.00   
6    Developer                                    [<class 'str'>]      100.00   

   nulos_%  
0     0.43  
1     0.00  
2     7.19  
3     0.50  
4     0.00  
5     0.00  
6     0.00  


Haciendo un analisis de que columnas voy a precisar de este dataset, por las funicones pedidas y el analisis de machine learning, identifique que la columna tags es redundante, por ende la puedo eliminar.

Tambien tengo que cambiar el nombre de titulo para que se refencie en las otras tablas.(Title por ItemName)

ReleaseDate  (nulos) =  2309 quedo con el tipo de dato de columna bien, y los valores faltantes quedaron como Nat(Not a Time)

In [241]:
# dropeamos Tags que no nos sirve
df_OutputSteamGames.drop(['Tags'], axis=1, inplace=True)

In [242]:
# Chequeamos como quedaron los nombres de las columnas en la tabla df_UserItems
df_UserItems.columns

Index(['ItemId', 'ItemName', 'PlaytimeForever', 'UserId', 'ItemsCount'], dtype='object')

In [243]:
# Vamos a cambiar en el nombre de la columna Title en df_outputGames para igualarlo con el nombre de la columna
# ItemName de df_UserItems
df_OutputSteamGames.rename(columns={'Title': 'ItemName'}, inplace=True)

In [244]:
# Ahora vamos a dropear los valores faltantes en la columna Genres
df_OutputSteamGames.dropna(subset=['Genres'], inplace=True)

In [245]:
# Chequeamos como quedo
df_OutputSteamGames['Genres'].isnull().sum()

0

In [246]:
# Volvemos a verificar como quedaron la columnas y los valores faltantes
df_info = verificar_tipo_datos(df_OutputSteamGames)
print(df_info)

Genres  (nulos) =  0
ItemName  (nulos) =  0
ReleaseDate  (nulos) =  2308
Price  (nulos) =  0
ItemId  (nulos) =  0
Developer  (nulos) =  0

filas completamente nulas:  0
  nombre_campo                                         tipo_datos  no_nulos_%  \
0       Genres                                   [<class 'list'>]      100.00   
1     ItemName                                    [<class 'str'>]      100.00   
2  ReleaseDate  [<class 'pandas._libs.tslibs.timestamps.Timest...       92.79   
3        Price                                  [<class 'float'>]      100.00   
4       ItemId                                    [<class 'int'>]      100.00   
5    Developer                                    [<class 'str'>]      100.00   

   nulos_%  
0     0.00  
1     0.00  
2     7.21  
3     0.00  
4     0.00  
5     0.00  


In [247]:
# Al final de la Transformacion nos quedamos con estos registros
num_rows, num_columns = df_OutputSteamGames.shape
print(f'Número de filas en el DataFrame: {num_rows} de 32135')

Número de filas en el DataFrame: 31994 de 32135


Ahora guardamos el Dataframe en un .csv ya con los procesos de transformacion realizados

In [248]:
# Definimos la ruta relativa
ruta_relativa = os.path.join('Datasets-Limpios', 'output_games.parquet')

# Guardamos el DataFrame en un archivo CSV con la ruta relativa
df_OutputSteamGames.to_parquet(ruta_relativa, index=False)

---------------------------------------------------------

Analizamos el Dataframe df_UserReviews

In [249]:
# vamos a verificar como esta el dataframe
df_info = verificar_tipo_datos(df_UserReviews)
print(df_info)

funny  (nulos) =  0
posted  (nulos) =  0
last_edited  (nulos) =  0
item_id  (nulos) =  0
helpful  (nulos) =  0
recommend  (nulos) =  0
review  (nulos) =  0
user_id  (nulos) =  0
user_url  (nulos) =  0

filas completamente nulas:  0
  nombre_campo        tipo_datos  no_nulos_%  nulos_%
0        funny   [<class 'str'>]       100.0      0.0
1       posted   [<class 'str'>]       100.0      0.0
2  last_edited   [<class 'str'>]       100.0      0.0
3      item_id   [<class 'str'>]       100.0      0.0
4      helpful   [<class 'str'>]       100.0      0.0
5    recommend  [<class 'bool'>]       100.0      0.0
6       review   [<class 'str'>]       100.0      0.0
7      user_id   [<class 'str'>]       100.0      0.0
8     user_url   [<class 'str'>]       100.0      0.0


In [250]:
# Chequeamos el dataframe
df_UserReviews.head()

funny                     posted last_edited item_id  \
0         Posted November 5, 2011.                1250   
1            Posted July 15, 2011.               22200   
2           Posted April 21, 2011.               43110   
3            Posted June 24, 2014.              251610   
4        Posted September 8, 2013.              227300   

                                           helpful  recommend  \
0                                   No ratings yet       True   
1                                   No ratings yet       True   
2                                   No ratings yet       True   
3  15 of 20 people (75%) found this review helpful       True   
4     0 of 1 people (0%) found this review helpful       True   

                                              review            user_id  \
0  Simple yet with great replayability. In my opi...  76561197970982479   
1               It's unique and worth a playthrough.  76561197970982479   
2  Great atmosphere. The gunplay can be a bit chu...  76561197970982479   
3  I know what you think when you see this title ...            js41637   
4  For a simple (it's actually not all that simpl...            js41637   

                                            user_url  
0  http://steamcommunity.com/profiles/76561197970...  
1  http://steamcommunity.com/profiles/76561197970...  
2  http://steamcommunity.com/profiles/76561197970...  
3               http://steamcommunity.com/id/js41637  
4               http://steamcommunity.com/id/js41637

Chequeamos los duplicados y los tratamos

In [251]:
# Duplicados
duplicates = df_UserReviews.duplicated()
print(f'Número total de filas duplicadas: {duplicates.sum()}')

Número total de filas duplicadas: 874


In [252]:
# Borro los duplicados
df_UserReviews.drop_duplicates(inplace=True)

In [253]:
# chequeamos como quedo
duplicates = df_UserReviews.duplicated()
print(f'Número total de filas duplicadas: {duplicates.sum()}')

Número total de filas duplicadas: 0


Ahora vamos a borrar las columnas que no nos sirven para el proyecto

In [254]:
# Vemos las columnas que tenemos
df_UserReviews.columns

Index(['funny', 'posted', 'last_edited', 'item_id', 'helpful', 'recommend',
       'review', 'user_id', 'user_url'],
      dtype='object')

In [255]:
# Vamos a borrar las columnas que no vamos a utilizar para el proyecto(last edited,  user_url, funny, helpful)
df_UserReviews.drop(columns= ['last_edited', 'user_url', 'funny','helpful'], inplace=True)

In [256]:
# chequeamos como quedo el dataframe
df_UserReviews.tail()

posted item_id  recommend  \
59300  Posted July 10.      70       True   
59301   Posted July 8.  362890       True   
59302   Posted July 3.  273110       True   
59303  Posted July 20.     730       True   
59304   Posted July 2.     440       True   

                                                  review            user_id  
59300  a must have classic from steam definitely wort...  76561198312638244  
59301  this game is a perfect remake of the original ...  76561198312638244  
59302  had so much fun plaing this and collecting res...        LydiaMorley  
59303                                                 :D        LydiaMorley  
59304                                     so much fun :D        LydiaMorley

Vamos a reorganizar y cambiar los nombres de las columnas

In [257]:
# Chequeamos cuales son los nombres del dataframe
df_UserReviews.columns

Index(['posted', 'item_id', 'recommend', 'review', 'user_id'], dtype='object')

In [258]:
# Vemos como quedaron las columnas con los nombres cambiados eb df_UserItems
df_UserItems.columns

Index(['ItemId', 'ItemName', 'PlaytimeForever', 'UserId', 'ItemsCount'], dtype='object')

Identificamos que ItemName de la Tabla df_UserItems es la misma que item_id de la tabla df_UserReviews, podemos usar el mismo nombre

In [259]:
# vamos a renombrar las columnas en una manera standard y siguiendo buenas practicas 
df_UserReviews = df_UserReviews.rename(columns={'posted': 'DatePosted'})
df_UserReviews = df_UserReviews.rename(columns={'item_id': 'ItemId'})
df_UserReviews = df_UserReviews.rename(columns={'recommend': 'Recommend'})
df_UserReviews = df_UserReviews.rename(columns={'review': 'Review'})
df_UserReviews = df_UserReviews.rename(columns={'user_id': 'UserId'})

In [260]:
# chequeamos como quedaron los cambios de nombres de las columnas
df_UserReviews.columns

Index(['DatePosted', 'ItemId', 'Recommend', 'Review', 'UserId'], dtype='object')

Chequeamos el tipo de datos de las columnas

In [261]:
df_UserReviews.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 58431 entries, 0 to 59304
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   DatePosted  58431 non-null  object
 1   ItemId      58431 non-null  object
 2   Recommend   58431 non-null  bool  
 3   Review      58431 non-null  object
 4   UserId      58431 non-null  object
dtypes: bool(1), object(4)
memory usage: 2.3+ MB


In [262]:
# Vamos a cambiar el tipo de datos para las columnas ItemId
df_UserReviews['ItemId'] = df_UserReviews['ItemId'].astype(int)

In [263]:
# ahora vemos el campo DatePosted que tiene un mal formato
df_UserReviews['DatePosted']

0         Posted November 5, 2011.
1            Posted July 15, 2011.
2           Posted April 21, 2011.
3            Posted June 24, 2014.
4        Posted September 8, 2013.
                   ...            
59300              Posted July 10.
59301               Posted July 8.
59302               Posted July 3.
59303              Posted July 20.
59304               Posted July 2.
Name: DatePosted, Length: 58431, dtype: object

In [264]:
# Ahora vamos a ver como podemos modificarlo para que nos quede en un formato correcto
#quitamos la palabra posted, las comas y puntos
df_UserReviews.DatePosted=df_UserReviews.DatePosted.str.removeprefix('Posted').str.replace(',','').str.replace('.','')

C:\Users\frank\AppData\Local\Temp\ipykernel_21320\2829926832.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_UserReviews.DatePosted=df_UserReviews.DatePosted.str.removeprefix('Posted').str.replace(',','').str.replace('.','')


In [265]:
# sacamos el año de la columna posted y creamos una nueva columna con el año solamente
df_UserReviews['YearPosted']=df_UserReviews.DatePosted.str.split(expand=True)[2]
df_UserReviews.head()

DatePosted  ItemId  Recommend  \
0    November 5 2011    1250       True   
1       July 15 2011   22200       True   
2      April 21 2011   43110       True   
3       June 24 2014  251610       True   
4   September 8 2013  227300       True   

                                              Review             UserId  \
0  Simple yet with great replayability. In my opi...  76561197970982479   
1               It's unique and worth a playthrough.  76561197970982479   
2  Great atmosphere. The gunplay can be a bit chu...  76561197970982479   
3  I know what you think when you see this title ...            js41637   
4  For a simple (it's actually not all that simpl...            js41637   

  YearPosted  
0       2011  
1       2011  
2       2011  
3       2014  
4       2013

In [266]:
# Ahora chequeamos cuantos nulos nos quedo en la nueva columna YearPosted
df_UserReviews['YearPosted'].isnull().sum()

9933

In [267]:
# Vamos a rellenar los años faltantes en el campo YearPosted con el año 2016 ya que es el año en curso en donde
# se hicieron los reviews y fue la fecha de Release del titulo
df_UserReviews.YearPosted.fillna('2016',inplace=True)

In [268]:
# Ahora chequeamos nuevamente como quedo el tratamiento que hicimos
df_UserReviews['YearPosted'].isnull().sum()

0

In [269]:
# Por ultimo vamos a borrar la columna en donde sacamos el año y ya no tiene utilidad
df_UserReviews.drop(columns='DatePosted', inplace=True)

In [270]:
# Para Terminar el proceso de transformacion en este dataframe vamos a cambiar el tipo de dato de YearPosted, que fue
# creado con el año que se creo la review, por int ya que representa un numero con el año solamente
df_UserReviews['YearPosted'] = df_UserReviews['YearPosted'].astype(int)

In [271]:
# Chequeamos como quedaron los tipos de datos en las columnas
df_UserReviews.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 58431 entries, 0 to 59304
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ItemId      58431 non-null  int32 
 1   Recommend   58431 non-null  bool  
 2   Review      58431 non-null  object
 3   UserId      58431 non-null  object
 4   YearPosted  58431 non-null  int32 
dtypes: bool(1), int32(2), object(2)
memory usage: 1.8+ MB


Por ultimo vamos a crear el .csv con la transformacion y limpieza hecha

In [272]:
# Definimos la ruta relativa
ruta_relativa = os.path.join('Datasets-Limpios', 'user_reviews.parquet')

# Guardamos el DataFrame en un archivo CSV con la ruta relativa
df_UserReviews.to_parquet(ruta_relativa, index=False)